File: IMPLEMENTATION/CacheSim.py
import random
import time

class CacheBlock:
    def __init__(self, block_size):
        self.block_size = block_size
        self.data = [None] * block_size
        self.tag = None
        self.valid = False
        self.timestamp = time.time()  # Add a timestamp field
        self.access_count = 0  # Add an access count field for LFU
        self.added_timestamp = time.time()  # Add an added timestamp field for FIFO

class CacheSet:
    def __init__(self, associativity, block_size):
        self.blocks = [CacheBlock(block_size) for _ in range(associativity)]
        
class Cache:
    def __init__(self, cache_size=1024, block_size=4, associativity=1, replacement_policy='Random'):
        self.num_sets = cache_size // (block_size * associativity)
        self.block_size = block_size
        self.associativity = associativity
        self.replacement_policy = replacement_policy
        self.sets = [CacheSet(associativity, block_size) for _ in range(self.num_sets)]
        self.access_count = 0
        self.hit_count = 0

    def access(self, operation, size, address):
        num_blocks = -(-size // self.block_size)  # Equivalent to math.ceil(size / self.block_size)
        for i in range(num_blocks):
            block_address = address + i * self.block_size
            self.access_block(operation, block_address)

    def access_block(self, operation, address):
        self.access_count += 1
        set_index = (address // self.block_size) % self.num_sets
        tag = address // (self.block_size * self.num_sets)
        cache_set = self.sets[set_index]

        for block in cache_set.blocks:
            if block.valid and block.tag == tag:
                self.hit_count += 1
                block.timestamp = time.time()
                block.access_count += 1
                return True
        
        for block in cache_set.blocks:
            if not block.valid:
                block.valid = True
                block.tag = tag
                block.data = self.fetch_data(operation,  address)
                block.timestamp = time.time()
                block.access_count = 1
                block.added_timestamp = time.time()
                return False

        if self.replacement_policy == 'Random':
            block = random.choice(cache_set.blocks)
        elif self.replacement_policy == 'LRU':
            block = min(cache_set.blocks, key=lambda block: block.timestamp)
        elif self.replacement_policy == 'FIFO':
            block = min(cache_set.blocks, key=lambda block: block.added_timestamp)
        elif self.replacement_policy == 'LFU':
            block = min(cache_set.blocks, key=lambda block: block.access_count)
        
        block.valid = True
        block.tag = tag
        block.data = self.fetch_data(operation,  address)
        block.timestamp = time.time()
        block.access_count = 1
        block.added_timestamp = time.time()
        return False

    def fetch_data(self, operation, address):
        if operation == 'IFetch':
            return f"Instruction at address {address}"
        elif operation in ['Read', 'Write']:
            return f'Data at address {address}'
        else:
            raise ValueError(f"Unknown operation: {operation}")


    def hit_rate(self):
        return self.hit_count / self.access_count if self.access_count > 0 else 0

    def parse_instructions(self, filename):
        with open(filename, 'r') as f:
            for line in f:
                operation, size, address = line.split()
                size = int(size)
                address = int(address, 16)
                self.access(operation, size, address)


cache_size = 1024
block_size = 4
associativity = 16
replacement_policy = 'FIFO'


def main():
    cache = Cache(cache_size, block_size, associativity, replacement_policy)
    print("╔═════════════════════════════════════════╗")
    print(f"║{'Cache total size: ' + str(cache_size) + ' bytes':^41}║")
    print(f"║{'Cache amount: ' + str(int(cache_size/block_size)):^41}║")
    print(f"║{'Block size: ' + str(block_size) + ' bytes':^41}║")
    print(f"║{'Number of sets: ' + str(associativity):^41}║")
    print(f"║{'Replacement policy: ' + replacement_policy:^41}║")
    print("╚═════════════════════════════════════════╝")
    cache.parse_instructions('mem.txt')
    print("╔═════════════════════════════════════════╗")
    print("║{:^50}║".format('\033[92m Simulation finished!\033[0m'))
    print(f"║{'Total access count: ' + str(cache.access_count):^41}║")
    print(f"║{'Total hit count: ' + str(cache.hit_count):^41}║")
    print(f"║{'Cache hit rate: ' + str(cache.hit_rate() * 100) + '%':^41}║")
    print("╚═════════════════════════════════════════╝")

if __name__ == "__main__":
    main()

File: FURTHER STEP/CacheSim.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <algorithm>

using namespace std;

class CacheBlock
{
public:
    int block_size;
    vector<int> data;
    int tag;
    bool valid;
    double timestamp;
    int access_count;
    double added_timestamp;
    int time;
    int add_time;

    CacheBlock(int block_size) : block_size(block_size), tag(-1), valid(false), timestamp(0.0), access_count(0), added_timestamp(0.0), time(0), add_time(0)
    {
        data.resize(block_size, -1);
    }
};

class CacheSet
{
public:
    vector<CacheBlock> blocks;

    CacheSet(int associativity, int block_size)
    {
        blocks.resize(associativity, CacheBlock(block_size));
    }
};

class Cache
{
public:
    int num_sets;
    int block_size;
    int associativity;
    string replacement_policy;
    vector<CacheSet> sets;
    int access_count;
    int hit_count;

    Cache(int cache_size = 1024, int block_size = 4, int associativity = 1, string replacement_policy = "Random") : block_size(block_size), associativity(associativity), replacement_policy(replacement_policy),
                                                                                                                    access_count(0), hit_count(0)
    {

        num_sets = cache_size / (block_size * associativity);
        sets.resize(num_sets, CacheSet(associativity, block_size));
    }

    void access(string operation, int size, int address)
    {
        int num_blocks = ceil((double)size / block_size);
        for (int i = 0; i < num_blocks; ++i)
        {
            int block_address = address + i * block_size;
            access_block(operation, block_address);
        }
    }

    void access_block(string operation, int address)
    {
        access_count++;
        int set_index = (address / block_size) % num_sets;
        int tag = address / (block_size * num_sets);
        CacheSet &cache_set = sets[set_index];

        for (CacheBlock &block : cache_set.blocks)
        {
            if (block.valid && block.tag == tag)
            {
                hit_count++;
                block.timestamp = ++block.time;
                block.access_count++;
                return;
            }
        }

        for (CacheBlock &block : cache_set.blocks)
        {
            if (!block.valid)
            {
                block.valid = true;
                block.tag = tag;
                block.data = fetch_data(operation, address);
                block.timestamp = ++block.time;
                block.access_count = 1;
                block.added_timestamp = ++block.add_time;
                return;
            }
        }

        CacheBlock &block = choose_block_to_replace(cache_set);
        block.valid = true;
        block.tag = tag;
        block.data = fetch_data(operation, address);
        block.timestamp = ++block.time;
        block.access_count = 1;
        block.added_timestamp = ++block.add_time;
    }

    vector<int> fetch_data(string operation, int address)
    {
        vector<int> data(block_size, -1);
        if (operation == "IFetch")
        {
            // cout << "Instruction at address " << address << endl;
        }
        else if (operation == "Read" || operation == "Write")
        {
            // cout << "Data at address " << address << endl;
        }
        else
        {
            // throw invalid_argument("Unknown operation: " + operation);
        }
        return data;
    }

    double hit_rate()
    {
        return (double)hit_count / access_count;
    }

    CacheBlock &choose_block_to_replace(CacheSet &cache_set)
    {
        if (replacement_policy == "Random")
        {
            return cache_set.blocks[rand() % associativity];
        }
        else if (replacement_policy == "LRU")
        {
            return *min_element(cache_set.blocks.begin(), cache_set.blocks.end(),
                                [](const CacheBlock &a, const CacheBlock &b)
                                {
                                    return a.timestamp < b.timestamp;
                                });
        }
        else if (replacement_policy == "FIFO")
        {
            return *min_element(cache_set.blocks.begin(), cache_set.blocks.end(),
                                [](const CacheBlock &a, const CacheBlock &b)
                                {
                                    return a.added_timestamp < b.added_timestamp;
                                });
        }
        else if (replacement_policy == "LFU")
        {
            return *min_element(cache_set.blocks.begin(), cache_set.blocks.end(),
                                [](const CacheBlock &a, const CacheBlock &b)
                                {
                                    return a.access_count < b.access_count;
                                });
        }
        else
        {
            throw invalid_argument("Unknown replacement policy: " + replacement_policy);
        }
    }

    void parse_instructions(string filename)
    {
        ifstream file(filename);
        if (!file.is_open())
        {
            cerr << "Error: Unable to open file " << filename << endl;
            return;
        }

        string operation;
        int size, address;
        while (file >> operation >> size >> hex >> address)
        {
            access(operation, size, address);
        }
        file.close();
    }
};

int main()
{
    srand(0); // 用于随机数生成的种子
    Cache cache(1024, 4, 16, "LFU");
    cache.parse_instructions("mem.txt");
    cout << "Simulation finished!" << endl;
    cout << "Total access count: " << cache.access_count << endl;
    cout << "Total hit count: " << cache.hit_count << endl;
    cout << "Cache hit rate: " << cache.hit_rate() * 100 << "%" << endl;
    return 0;
}
